snuba/utils/streams/batching.py

```diff
@@ -74,25 +80,55 @@ class Batch(Generic[TResult]):
     processing_time_ms: float = 0.0
 
 
-class BatchProcessor:
+class BatchProcessorFactory(ProcessorFactory[TPayload]):
```

Comment:
I am not sure I get the value of having a `ProcessorFactory` class. It seems to me there would be a 1:1 relationship between Processor class and ProcessorFactory (or do you expect to have more generic factories that produce multiple Processor types ? ).
If not, can't this be a Protocol or a simple partially applied function?

Reply:
> I am not sure I get the value of having a `ProcessorFactory` class. It seems to me there would be a 1:1 relationship between Processor class and ProcessorFactory (or do you expect to have more generic factories that produce multiple Processor types ? ).

I think this is probably the right way to look at it. I don't have anything more complex planned (at least right now, that might change later, I'm not sure.)

> If not, can't this be a Protocol or a simple partially applied function?

It could be. I think there is value in having a named type to refer to and subtype, etc. but not enough necessarily enough value to justify the overhead here. I'll look at simplifying this.

Reply:
So, I tried this with 2ba5922d0fb0b24fbbfdce2e4c95354a1466fd9e:

```
(.venv) ted@veneno % python -m mypy --strict snuba/cli/replacer.py
[…]
snuba/cli/replacer.py:118: error: Argument 3 to "StreamProcessor" has incompatible type "partial[BatchProcessor[TPayload]]"; expected "Callable[[Callable[[Mapping[Partition, int]], None]], Processor[KafkaPayload]]"
snuba/cli/replacer.py:118: note: "partial[BatchProcessor[TPayload]].__call__" has type "Callable[[VarArg(Any), KwArg(Any)], BatchProcessor[TPayload]]"
[…]
```

Since ad-hoc partials aren't really a good solution from a typing perspective, the only way that I can think to get around this would be for each processor (soon to be "processing strategy") to define a function that generally looks the same as the factory class but without the instance variables. At that point, I think that the definition of the `ProcessorFactory` class (soon to be `ProcessingStrategyFactory`) is clearer than requiring each implementation to define a factory function that matches `Callable[[Callable[[Mapping[Partition, int]], None]], Processor[TPayload]]`.

With that in mind, I'm tempted to leave this as-is (except for the naming.) Do you have any issues with that or other ideas?

Reply:
well, this is annoying. I think declaring a dedicated factory function would not be worse than than the Factory class, so I think I am ok with the factory.
What about instead a factory class method in the processor itself? I think python allows us to have an abstract class method.

Reply:
Hm, mypy doesn't seem to like that either:

```diff
diff --git a/snuba/utils/streams/batching.py b/snuba/utils/streams/batching.py
index e0105155..65dcf8ba 100644
--- a/snuba/utils/streams/batching.py
+++ b/snuba/utils/streams/batching.py
@@ -240,3 +240,15 @@ class BatchProcessor(ProcessingStrategy[TPayload]):
         logger.debug("Offset commit took %dms", commit_duration)
 
         self.__batch = None
+
+    @classmethod
+    def factory(
+        cls,
+        worker: AbstractBatchWorker[TPayload, TResult],
+        max_batch_size: int,
+        max_batch_time: int,
+        metrics: MetricsBackend,
+    ) -> Callable[
+        [Callable[[Mapping[Partition, int]], None]], ProcessingStrategy[TPayload]
+    ]:
+        raise NotImplementedError
diff --git a/snuba/utils/streams/processing.py b/snuba/utils/streams/processing.py
index d28c55b6..24a189e1 100644
--- a/snuba/utils/streams/processing.py
+++ b/snuba/utils/streams/processing.py
@@ -35,6 +27,14 @@ class ProcessingStrategy(ABC, Generic[TPayload]):
     def close(self) -> None:
         raise NotImplementedError
 
+    @abstractclassmethod
+    def factory(
+        cls,
+    ) -> Callable[
+        [Callable[[Mapping[Partition, int]], None]], ProcessingStrategy[TPayload]
+    ]:
+        raise NotImplementedError
+
 
 class StreamProcessor(Generic[TPayload]):
     def __init__(
```

This results in:

```
snuba/utils/streams/batching.py:245: error: Signature of "factory" incompatible with supertype "ProcessingStrategy"
```

Reply:
oh well, yes ... without variadics on the parent class it would not work.  
At this point, no strong opinion between the factory class and the variable number of arguments.

Reply:
I think I've seen mypy allow adding additional arguments to a function defined on the supertype, but I don't think that was on an abstract class (or I'm just misremembering entirely.)

For now, I've at least updated the naming to bring it more inline with the other usages. I don't love the factory pattern either, but it at least works and we can come back and change it later if we'd like to.